server/lib/pbench/process_tb.py

```diff
@@ -76,61 +94,305 @@ def _results_push(controller: str, tb: Path, token: str):
+        """Create destination symlinks for state directories.
+
+        The `TODO` state is skipped entirely, and the list of dispatch states
+        are used directly.
+
+        Returns:
+            0 on success, a count of errors on failure
+        """
+        target = dest / tb.name
+        if not target.is_file():
+            self.logger.error(
+                "{}: INTERNAL ERROR - destination tar ball {} does not exist",
+                self.config.TS,
+                target,
+            )
+            return 1
+        nerr = 0
+        for state_d_str in self.dispatch_states:
+            state_d = dest / state_d_str
+            state_d_tb = state_d / tb.name
+            try:
+                state_d_tb.symlink_to(target)
+            except Exception as exc:
+                self.logger.error(
+                    "{}: Cannot create {} link to {}: '{}'",
+                    self.config.TS,
+                    state_d_tb,
+                    target,
+                    exc,
+                )
+                nerr += 1
+        return nerr
+
     def process_tb(self) -> Results:
-        """Searches for tarballs in the configured receive directory and
-        uploads them to the Pbench server"""
+        """Searches for tarballs in the configured receive directory and uploads
+        them to the Pbench server
+
+        Returns:
+            a Results named tuple
+        """
 
         # Check for results that are ready for processing: version 002 agents
         # upload the MD5 file as xxx.md5.check and they rename it to xxx.md5
         # after they are done with MD5 checking so that's what we look for.
         self.logger.info("{}", self.config.TS)
         nstatus = ""
-        ntotal = ntbs = nerr = 0
+        ntotal = ntbs = ndups = nqua = nerr = 0
 
         for tbmd5 in sorted(self.receive_dir.glob("**/*.tar.xz.md5")):
             ntotal += 1
 
-            # Extracts full pathname of tarball from the name of hashfile
-            # by trimming .md5 suffix
+            # Old prep-shim did (quarantining any failed step)
+            # 1. check if archive tree already has the tar ball or .md5 in it
+            #    and quarantine if so
+            # 2. create the TODO sub-directory
+            # 3. Copy the tar ball .md5 file to the archive tree
+            # 4. Move the tar ball file to the archive tree
+            # 5. restorecon in the archive tree destination
+            # 6. remove the reception area .md5 file
+            # 7. create a symlink for the tar ball in the TODO directory
+
+            # Old dispatch did (quarantining "DUPLICATE__NAME" tar balls, etc.)
+            # 1. Creates all the state directories
+            # 2. Verifies the tar ball's MD5 signature
+            # 3. Creates all the link destination directory symlinks
+
+            # Combined operations will do
+            # 1. PUT API if have a token
+            # 2. p1
```

Comment:
I hope not: if we're operating in "prep-shim" mode with shared `ARCHIVE` then at p1 we *expect* that `PUT` has already populated the `ARCHIVE` tree and that's *good*, not *bad*; in fact, it's an error if the tarball isn't there. 😆 

Actually, since you seem to be building the new prep-shim to operate in at least two of three possible modes, it may be that trying to express the flow in a single list isn't going to be practical.